File: 백준/Silver/9012. 괄호/괄호.py
def is_balanced(line):
    stack = []

    for ch in line:

        if ch == '(' or ch == '[' or len(stack) == 0:
            stack.append(ch)
        
        elif ch == ')':
            if stack[-1] == '(':
                stack.pop()
            else:
                stack.append(ch)

        elif ch == ']':
            if stack[-1] == '[':
                stack.pop()
            else:
                stack.append(ch)

    return len(stack) == 0


n = int(input())

for i in range(n):
    line = input()

    if is_balanced(line):
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.cc
#include<iostream>
#include<string>
#include<algorithm>
#include<vector>
#include<stack>

using namespace std;

string str;

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL), cout.tie(NULL);

	int N;
	cin >> N;
	
	for (int i = 0; i < N; i++)
	{
		stack<char>s;
		cin >> str;
		for (int j = 0; j < str.size(); j++)
		{
			if (s.empty())
			{
				if (str[j] == '(')
					s.push(str[j]);
				else
				{
					s.push('1');
					break;
				}
			}
			else
			{
				if (str[j] == '(')
					s.push(str[j]);
				else
				{	
					s.pop();
				}
			}
		}
		if (s.empty())
			cout << "YES\n";
		else
			cout << "NO\n";

	}
	

	return 0;
}